maven_iuvs/echelle.py

```diff
@@ -825,3 +966,1289 @@ def find_files_with_geometry(file_index):
+    function of dlambda = x-lambda_c, where x is the observational data
+    and LSF_x is the x points defined for the instrument LSF. This is because
+    the instrument LSF is defined by delta lambda, the difference from the 
+    central wavelength of the emission. Finally, the computed CLSF is inteprolated
+    at the bin edges. 
+
+    TODO: This function will eventually have to be modified to accept the bin edges from the
+    actual instrument rather than just making it up, but currently the recorded bin widths 
+    are for the standard resolution mode, not the echelle mode. 
+    
+    Parameters
+    ----------
+    lambda_c : float
+               a single wavelength in nm, candidate for a possible central wavelength marking the location of the emission peak.
+    bin_edges : array
+                Defines the bin edges in nm on which the LSF will be interpolated
+    CLSF : array
+           CLSF computed from the LSF in question, result of CLSF_from_LSF.
+
+    Returns
+    ----------
+    interp_CLSF : array
+                  the same CLSF reinterpolated on bin_edges.
+    """
+  
+    # Shift the wavelengths of the bin centers and edges
+    dlambda_binedges = binedges - lambda_c
+    
+    # Ensure that CLSF x is increasing everywhere so interp doesn't have meaningless results
+    if any(np.diff(CLSF) < 0):
+        raise Exception("ValueError: Can't interpolate because x values are not monotonically increasing")
+
+    interp_CLSF = np.interp(dlambda_binedges, CLSF[:, 0], CLSF[:, 1], left=0, right=1) 
+
+    # For some reason, interp function isn't automatically setting the edges to the requested values
+    if any(interp_CLSF > 1):
+        interp_CLSF[np.where(interp_CLSF > 1)] = 1
+
+    return interp_CLSF
+    
+
+def CLSF_from_LSF(LSFx, LSFy):
+    """
+    Compute the empirical CLSF, given some LSF defined by xdata and ydata.
+    
+    Parameters
+    ----------
+    LSFy, LSFx : arrays
+                 x and y points which define the instrument LSF.
+    dx : real (float or int)
+         difference between consecutive x values. If not provided, will be calculated from xdata.
+
+    Returns
+    ----------
+    cumulative : array
+                 The CLSF, valid at the same xdata.
+    """
+
+    # The LSF usually comes in angstroms, and covers ± ~3 Å; put it in nm.
+    if 1 <= np.max(LSFx)<= 4:
+        m = (1/10)
+    elif np.max(LSFx) < 1:
+        m = 1
+
+
+    cumulative = np.zeros((len(LSFx), 2))
+    
+    # Fill the wavelengths...
+    cumulative[:, 0] = LSFx * m
+
+    # Now fill the cumulative flux at each wavelength. Bceause the value is specified
+    # in the middle of the wavelength bin, the cumulative amount at each bin center is
+    # half the value at the previous bin center plus half the value at the present bin center.
+    incr_array = 0.5 * (LSFy[:-1] + LSFy[1:])
+    incr_array = np.insert(incr_array, 0, LSFy[0]/2) # At the first entry, the cumulative amount is 0.
```

Comment:
comment should read "At the first entry (halfway through the first bin), cumulative brightness is half the brightness of the first bin"

Reply:
Yes, must have transposed the 0 index into a value in brain

Reply:
Fixed by: f0c60d192fd4cfebf083e2c2a9930d1f48e90bcf